SPEC.md

```diff
@@ -0,0 +1,121 @@
+# CNI Specification
+Note that the prime source of information about CNI is to be considered the reference implementation and test suite.
+This document is a reading companion (rather than the source of truth) that should help expediate the development of implementations.
+
+## Core Language
+A CNI document consists of an arbitrary amount of statements.
+A statement may either be a section declaration, key-value pair, or key-rawvalue pair.
+
+### Key
+A key is composed of alphanumeric (`[a-zA-Z0-9]`) characters, underscores, and dashes.
+Keys may contain dots, but may not start or end with them.
+Two dots in a row are disallowed (due to being nonsensical), but implementations do not need to explicitly disallow them.
+
+### Section Declaration
+A section consists of an opening bracket (`[`), the section name, and a closing bracket (`]`).
+The section name may either be a valid key, or empty.
+
+Until the next section, all keys will be automatically prepended the section name and a `.`, or with nothing (if the section name is the empty string).
+
+### Key-(Raw)Value Pairs
+A key-value and key-rawvalue pair look very similar.
+It's a key, an equals, and a value, with arbitrary amounts of spaces before and after the `=`.
+
+A value may contain any UTF-8 characters except vertical whitespace or the `#` sign (note: this is expanded upon in ini-compatibility).
+It may not start or end with whitespace, or start with a backtick.
+
+A rawvalue starts with a backtick and ends with a backtick, but may contain any UTF-8 characters besides those.
+If a backtick needs to be inserted into a raw value, a backtick may be escaped with a second one.
+
+Theoretically, this means that both values and raw values may contain arbitrary bytes, but the specification is exclusively intended to be used with UTF-8 data.
+Implementations are not required to validate it, though.
+
+### Comments
+A comment starts with a `#` (note: this is expanded upon in ini-compatibility) and ends at the next newline.
+A comment may start on its own line, at any point inside of a value, after a section declaration, or after a raw value.
+It is impossible to have comments inside of a raw value, or inside of a section declaration.
+
+## Ini-Compatibility
+For compatibility with INI, the `;` character may also be used for comments.
+In this case, anytime the `#` is a significant character in the core language, the `;` character should be as well.
+
+## Extensions
+All extensions are fully optional.
+Currently, multiple of them are consequences of keeping the grammar lean, and are documented, but only parts of their behaviors are desirable.
```

Comment:
I understand this as "fully optional for libraries to implement", while from #2 I gather it should mean fully optional to library implementers as well as library users.
```suggestion
All extensions are fully optional to be implemented. Implementations should allow for specific extensions to be en- or disabled.
Currently, multiple of them are consequences of keeping the grammar lean (so you will probably find it easier to implement them than not), and are documented, but only parts of their behaviors are desirable.
```

Reply:
When I say "users" I mean end-users (application users). I.e the people writing the CNI files.
Library users have the option on the basis of the library they end up picking.

The intended workflow is:
1. You realize you're working with a CNI file, you know what CNI is (`core`) and can write it.
2. If the application author tells you to use an extension (or you realize that an extension is available), you know you can take advantage of its features for that application only.

Basically, `ext` is for stuff that I don't want CNI as a whole to be associated with / preassumed availability of.
`ini` can in fact be considered to be `ext/ini`, but it's separated out (right now) for convenience purposes.